keps/sig-cli/2906-kustomize-function-catalog/README.md

```diff
@@ -0,0 +1,667 @@
+* https://github.com/features/packages
+* https://github.com/goharbor/harbor/releases/tag/v2.0.0
+* https://github.com/distribution/distribution
+
+This proposal does not suggest adding any OCI artifact publishing capabilities to Kustomize, and would instead rely upon the [ORAS](https://oras.land) project to handle publishing and fetching of artifacts for now. Built in capabilities could be added by including ORAS as a library, but an analysis of the dependencies introduced will be needed.
+
+### Risks and Mitigations
+
+This proposal introduces extension capabilities to Kustomize that may expose users to external content. As with `Composition`, it must be made clear to users that use of a `Catalog` may represent untrusted/unvalidated content and they should only use `Catalogs` that they trust. When `Catalog` and other resources are stored as OCI artifacts, users can get extra assurance of content by using `digest` references. Additionally, the [cosign](https://github.com/sigstore/cosign) project could be used to provide signing and validation capabilities. The guidance around executing plugins, as outlined in the [Composition](../2290-kustomize-plugin-composition/README.md) KEP remain applicable when combined with `Catalog` resources.
+
+## Design Details
+
+The catalog kind will have a YAML representation. This representation will contain metadata about the catalog, such as name labels, as well as a collection of plugins. Each plugin entry will contain an apiVersion and kind, along with one or more references to a plugin provider, as well as an optional information, such as Open API v3 definitions. Provider references can contain https, git, or OCI references. Additionally, a provider can declare that it requires the use of network or storage mounts, which would otherwise be prohibited. The use of these requires [additional flags](https://github.com/kubernetes-sigs/kustomize/blob/1e1b9b484a836714b57b25c2cd47dda1780610e7/api/types/pluginrestrictions.go#L51-L55) on the command line. 
+
+When using OCI references, either a tag or digest reference can be provided. Exec plugins should include an sha256 hash for verification purposes, although this can also be done by using an OCI digest reference. When a hash verification fails, Kustomize will emit an error to inform the user. When a hash is not provided for verification, Kustomize will emit a warning to inform the user that validation could not be performed. 
+
+A complete representation is shown below:
+
+```yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-plugins"
+  labels:
+    author: ExampleCo
+spec: 
+  modules: 
+  - apiVersion: example.com/v1
+    kind: JavaApplication
+    description: "A Kustomize plugin provider that can handle Java apps"
+    definition: "https://example.com/java/definition"
+    provider: 
+      container: 
+        image: example/module_providers/java:v1.0.0
+        requireNetwork: true
+        requireFilesystem: true
```

Comment:
Do we cover how filesystem access will be done ?

Reply:
Why does this need to be different from `network` and `mounts` fields currently supported in the container function spec? https://github.com/kubernetes-sigs/kustomize/blob/f122fb12f3d0d234564b664c48fdb8d93ba8a7c6/kyaml/fn/runtime/runtimeutil/functiontypes.go#L149

Is requiring access to the entire filesystem necessary vs mounting the necessary files into the container?

Reply:
I'll reconcile this with the `functiontypes.go` struct. I think it will need to deviate a little bit from the `StorageMounts` field.

I think there are two aspects to this: 

1.) the catalog needs to express that yes, this function needs to access files
2.) these are the destinations that need to be populated

Within a kustomization (as far as I understand) it's going to mount within the kustomization root, not arbitrary files or the entire filesystem.  So we could leverage the existing `StorageMount` type, but omit the src elements, as those will be determined by the user of the function + the kustomization, not the catalog. 
So perhaps
```

wdyt?

Reply:
> Within a kustomization (as far as I understand) it's going to mount within the kustomization root, not arbitrary files or the entire filesystem.

That's what is intended but I actually don't think that's how it works today. I think we need to fix the code to ensure that the files mounted into the container are (a) within the kustomization root and (b) relative to the kustomization that refers to them - but this is an orthogonal issue that can be addressed separately.

> So we could leverage the existing StorageMount type, but omit the src elements, as those will be determined by the user of the function + the kustomization, not the catalog.

That sounds reasonable to me. Do you have an idea for what the syntax should look like in a `Kustomization` or `Composition` to specify `mounts.src`?